docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,574 @@
+  parameters:
+    trustPolicies:
+      - name: build-test-verification
+        keys:
+          - kms: inline-key-build
+          - kms: inline-key-test
+        enforcement: all
+```
+- Bob attempts to deploy a pod from the vetted image that has 1 cosign signature, which is signed with key in KeyManagementSystem `inline-key-build`. Pod FAILS verification and blocked.
+- Bob attempts to deploy a pod from the vetted image that has 2 cosign signatures, which is signed with keys in KeyManagementSystem `inline-key-build` & `inline-key-test`. Pod passes verification and is created.
+
+> **NOTE**: Based on Cosign's own verification implementation, there does not seem to be a way to enforce that ALL signatures found are valid. This behavior is unchartered as far as we can tell and would require Ratify to determine if we want to support this directly. For that reason, the strict ALL signature scenario is not yet included.
+
+#### Extra Considerations
+- How does Kyverno handle multiple Keys?
+  - Each key specified is a separate cosign verification operation
+  - User specifies enforcement type by specifying `count` of verification operations that must succeed.
+    - e.g 2 keys specified and 2 signatures on the image. each verification is against both signatures for a single key
+    - If count is 2, then 2 verifications must return true
+- How will this work with OCI 1.1 enabled cosign?
+  - Cosign OCI 1.1 support is experimental
+  - Currently, an arbitrary signature from list of referrers is chosen to perform verifications on IF multiple cosign artifacts are found.
+  - This logic will be updated in the future once 1.1 is GA but for now it's left as a TODO in the Cosign code base.
+
+### Implementation Details
+
+#### Handling multiple verifications at once
+- Ratify will largely mirror how cosign performs multi-signature verification
+  - cannot invoke cosign's multi signature verification API since the API exposed also includes the registry content fetch which is independent in Ratify
+- separate go routines per signature and collect all results PER KEY 
+
+## Support RSA and ED25519 keys for verification
+
+Each key specified will encapsulate a single verification context for all signatures found in the artifact that the verifier is running for.
+
+Verification flow per key:
+- Take raw key content (byte array) and [`LoadPublicKeyRaw()`](https://github.com/sigstore/cosign/blob/daf1eeb5ff2022d11466b8eccda25768b2dd2992/pkg/signature/keys.go#L91C1-L91C1)
+- The `LoadPublicKeyRaw()` method in the cosign keys library will automatically detect which key type (based on the key header) and return the appropriate signature verifier
+- Verifier is then passed to cosign's `VerifySignature()` method as a verification option
+
+## Dev Work Items (WIP)
+- Introduce new `KeyManagementSystem` CRD to replace `CertificateStore` (~ 3 weeks)
+  - maintain old `CertificateStore` controllers and source code for backwards compat
```

Comment:
it sounds great that we have backward compat for certificateStore. Will we consider add one warning for this crd is deprecating?

Reply:
Yes we will add deprecation warning here. We will continue to support all the way til Ratify become v2.0 but we will not add features to CertificateStore. And CertificateStore and KeyManagementSystem cannot be used together. We will implement logic in controller to throw error and show error in status